presto-main-base/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -714,8 +764,28 @@ private boolean retryConditionsMet(QueryResults queryResults)
         }
 
         if (!retryQueryWithHistoryBasedOptimizationEnabled(session)) {
-            if (!queryResults.getError().isRetriable()) {
-                return false;
+            // Check if cross-cluster retry is attempted
+            if (retryUrl.isPresent()) {
+                // Check if this is already a retry query - prevent cross-cluster retry chains
+                if (isRetryQuery) {
+                    log.debug("Query %s is already a retry query, preventing cross-cluster retry chain", queryId);
+                    return false;
+                }
```

Comment:
Since this is going to be set only on the backup cluster query, can this check be outside?

Reply:
I think it needs to be within the outer if statement, because we should never allow a cross-cluster retry on a query which itself as posted as a retry query on a backup cluster.